File: other/problem18.py
def solve():
    # Читаем треугольник из файла
    with open('Lab1/resources/0018_triangle.txt', 'r') as file:
        triangle = [[int(num) for num in line.strip().split()] for line in file]
    
    # Копируем последнюю строку треугольника
    dp = triangle[-1].copy()
    
    # Идем снизу вверх
    for row in range(len(triangle) - 2, -1, -1):
        new_row = [0] * (row + 1)
        for col in range(row + 1):
            # Для каждого элемента выбираем максимальный путь
            new_row[col] = triangle[row][col] + max(dp[col], dp[col + 1])
        dp = new_row
    
    return dp[0]

if __name__ == "__main__":
    print(solve())

File: src/FSharpLab1/Problem18.fs
module Problem18

open System.IO

let trianglePath = 
    Path.Combine(__SOURCE_DIRECTORY__, "..", "..", "resources", "0018_triangle.txt")

let getTriangle path =
    File.ReadAllLines path
    |> Array.map (fun line -> 
        line.Split(' ') 
        |> Array.map int)

let recursive () =
    let triangle = getTriangle trianglePath
    
    let rec findMax row col =
        if row >= Array.length triangle then 0
        else
            let current = triangle.[row].[col]
            let left = findMax (row + 1) col
            let right = findMax (row + 1) (col + 1)
            current + max left right
            
    findMax 0 0

let dynamic () =
    let triangle = getTriangle trianglePath
    let mutable dp = Array.copy (triangle.[Array.length triangle - 1])
    
    for row = (Array.length triangle - 2) downto 0 do
        let mutable newRow = Array.zeroCreate (row + 1)
        for col = 0 to row do
            newRow.[col] <- triangle.[row].[col] + max dp.[col] dp.[col + 1]
        dp <- newRow
        
    dp.[0]

let functional () =
    let triangle = getTriangle trianglePath
    
    triangle
    |> Array.rev
    |> Array.reduce (fun lower upper ->
        Array.init (Array.length upper) (fun i ->
            upper.[i] + max lower.[i] lower.[i + 1]))
    |> Array.head

let sequence () =
    let triangle = getTriangle trianglePath
    
    let folder currentRow previousRow =
        Seq.zip3 
            (Seq.append currentRow (seq { yield 0 }))
            (Seq.append (seq { yield 0 }) currentRow)
            previousRow
        |> Seq.map (fun (left, right, value) -> max left right + value)
        |> Seq.toArray
        
    triangle
    |> Array.rev
    |> Array.reduce folder
    |> Array.head

let solutions = [ recursive; dynamic; functional; sequence ]